modules/cmdb/endpoints/permission.go

```diff
@@ -231,102 +230,20 @@ func (e *Endpoints) getPermission(userID string, scopeType apistructs.ScopeType,
-		Access:           false,
-		PermissionList:   make([]apistructs.ScopeResource, 0),
-		ResourceRoleList: make([]apistructs.ScopeResource, 0),
-		Roles:            make([]string, 0),
-		Exist:            true,
-	}
-
-	// 若为系统管理员 & 查询系统范围权限，则返回true；若系统管理员查询企业/项目/应用等，应返回false
-	if e.member.IsAdmin(userID) && scopeType == apistructs.SysScope {
-		permissionList.Access = true
-		permissionList.Roles = []string{types.RoleSysManager}
-		return permissionList, nil
-	}
-
-	var roles []string
-	if userID != apistructs.SupportID {
-		members, err := e.member.GetByUserAndScope(userID, scopeType, scopeID)
-		if err != nil {
-			logrus.Infof("failed to get permission, (%v)", err)
-			return permissionList, errors.Errorf("failed to access permission")
-		}
 
-		if len(members) == 0 {
-			if scopeType == apistructs.SysScope {
-				return permissionList, nil
-			}
-			isPublic, err := e.permission.CheckPublicScope(userID, scopeType, scopeID)
-			if err != nil || !isPublic {
-				return permissionList, err
-			}
-			roles = append(roles, types.RoleGuest)
-		}
-
-		for _, member := range members {
-			if member.ResourceKey == apistructs.RoleResourceKey {
-				roles = append(roles, member.ResourceValue)
-			}
-		}
-		if len(roles) == 0 {
-			logrus.Warningf("nil role, scope: %s, scopeID: %d, memberInfo: %v", scopeType, scopeID, members)
-			return permissionList, nil
-		}
-	} else {
-		if scopeType == apistructs.SysScope {
-			return permissionList, nil
-		}
-		roles = []string{types.RoleOrgSupport}
-	}
-
-	var pml = make([]model.RolePermission, 0)
-	// get permissions from db
-	pmlDb, resourceRoleDb := e.db.GetPermissionList(roles)
-
-	// get permissions from permission.yml
-	pmlYml, resourceRoles := conf.RolePermissions(roles)
-	if len(pmlDb) > 0 {
-		for _, v := range pmlDb {
-			k := strutil.Concat(v.Scope, v.Resource, v.Action)
-			pmlYml[k] = v
-		}
-	}
-
-	// merge permission list
-	for _, v := range pmlYml {
-		pml = append(pml, v)
-	}
-
-	permissions := make([]apistructs.ScopeResource, 0)
-	for _, v := range pml {
-		permission := apistructs.ScopeResource{
-			Resource:     v.Resource,
-			Action:       v.Action,
-			ResourceRole: v.ResourceRole,
-		}
-		permissions = append(permissions, permission)
-	}
-
-	resourceRoles = append(resourceRoles, resourceRoleDb...)
-	resourceRoleList := make([]apistructs.ScopeResource, 0)
-	for _, v := range resourceRoles {
-		rr := apistructs.ScopeResource{
-			Resource:     v.Resource,
-			Action:       v.Action,
-			ResourceRole: v.ResourceRole,
-		}
-		resourceRoleList = append(resourceRoleList, rr)
-	}
+	permissionAdaptor.Once.Do(func() {
+		permissionAdaptor.Db = e.db
+		permissionAdaptor.Bdl = e.bdl
+		permissionAdaptor.RegisterHandler(security.AdminUserPermissionHandler{Adaptor: &permissionAdaptor})
+		permissionAdaptor.RegisterHandler(security.SupportUserPermissionHandler{Adaptor: &permissionAdaptor})
+		permissionAdaptor.RegisterHandler(security.NormalUserPermissionHandler{Adaptor: &permissionAdaptor})
+	})
```

Comment:
is it better initialize security.PermissionAdaptor as service in initialize.go? cc @sfwn

Reply:
it is found that the two processing logics of check and access are somewhat different, so the permission processor in the adaptor is dynamically configurable for each interface

You can find two places use adapter, and their number of processors is different

Reply:
i know then, but it can still be initialized as two service?